File: MyLeetCode/python/Combination Sum IV.py
###
# DP
# Time Complexity: O(nlogn + target*n)
# Space Complexity: O(target)
# State: dp[i] number of ways that can sum to i
# Function: dp[i] += dp[j-num] for num in nums
# Initialization: dp[0] = 1
# Answer: dp[target]
###
class Solution(object):
    def combinationSum4(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if not nums:
            return 0
        dp = [0] * (target + 1)
        nums.sort()
        dp[0] = 1
        for i in xrange(1, target + 1):
            for e in nums:
                if e <= i:
                    dp[i] += dp[i-e]
                else:
                    break
        return dp[target]

File: MyLeetCode/Leetcode/Combination Sum IV.cpp
/* DP
 * consider inserting the num as the last number
 */
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        int n = nums.size();
        if (n == 0) return 0;
        vector<int> dp(target+1);
        sort(nums.begin(), nums.end());
        dp[0] = 1;
        for (int i = 1; i <= target; i++) {
            for (int j = 0; j < n && nums[j] <= i; j++) {
                dp[i] += dp[i - nums[j]];
            }
        }
        return dp[target];
    }
};
 
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<int> res(target + 1);
        for (int i = 1; i <= target; i++) {
            for (int n : nums) {
                if (n > i) break;
                else if (n == i) {
                    res[i]++;
                }
                else{
                    res[i] += res[i - n];
                }
            }
        }
        return res[target];
    }
};